### PR TITLE
Estate variations endpoint

### PR DESCRIPTION
**Pull request by JoshuaH01**

### CODE REVIEW

File: app/uk/gov/hmrc/trusts/controllers/EstateVariationsController.scala [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+                                      desService: DesService,
+                                      auditService: AuditService
+                                    ) extends EstateBaseController with ValidationUtil {
+
+  def estateVariation() = identify.async(parse.json) {
+    implicit request =>
+
+      request.body.validate[EstateVariation].fold(
+        errors => {
+          Logger.error(s"[variations] trusts validation errors from request body $errors.")
```
Comment (jassalrichy) [inline app/uk/gov/hmrc/trusts/controllers/EstateVariationsController.scala]: rename "trusts validation errors" to "estate validation errors"

File: test/uk/gov/hmrc/trusts/connectors/DesConnectorSpec.scala [hunk 2]
```diff
@@ -154,6 +154,7 @@ class DesConnectorSpec extends BaseConnectorSpec {
     }
   }//trustsmatch
 
+
   ".checkExistingEstate" should {
 
     "return Matched " when {
```
Comment (jassalrichy) [inline test/uk/gov/hmrc/trusts/connectors/DesConnectorSpec.scala]: "trusts data match with existing trusts" to "estate data match with existing estate"

File: test/uk/gov/hmrc/trusts/controllers/EstateVariationsControllerSpec.scala [hunk 3]
```diff
@@ -31,56 +29,55 @@ import uk.gov.hmrc.trusts.BaseSpec
...
 
-        when(mockDesService.variation(any[Variation])(any[HeaderCarrier]))
+        when(mockDesService.estateVariation(any[EstateVariation])(any[HeaderCarrier]))
           .thenReturn(Future.successful(VariationResponse(tvnResponse)))
 
         when(mockConfig.auditingEnabled).thenReturn(false)
-        when(mockConfig.variationsApiSchema).thenReturn(appConfig.variationsApiSchema)
 
-        val requestPayLoad = Json.parse(validVariationsRequestJson)
+        val requestPayLoad = Json.parse(validTrustVariationsRequestJson)
```
Comment (jassalrichy) [inline test/uk/gov/hmrc/trusts/controllers/EstateVariationsControllerSpec.scala]: should this be "validEstateVariationsRequestJson" ?

File: test/uk/gov/hmrc/trusts/controllers/EstateVariationsControllerSpec.scala [hunk 4]
```diff
@@ -90,22 +87,21 @@ class VariationsControllerSpec extends BaseSpec with BeforeAndAfter {
...
 
-        when(mockDesService.variation(any[Variation])(any[HeaderCarrier]))
+        when(mockDesService.estateVariation(any[EstateVariation])(any[HeaderCarrier]))
           .thenReturn(Future.successful(VariationResponse(tvnResponse)))
 
         when(mockConfig.auditingEnabled).thenReturn(true)
-        when(mockConfig.variationsApiSchema).thenReturn(appConfig.variationsApiSchema)
 
-        val requestPayLoad = Json.parse(validVariationsRequestJson)
+        val requestPayLoad = Json.parse(validTrustVariationsRequestJson)
```
Comment (jassalrichy) [inline test/uk/gov/hmrc/trusts/controllers/EstateVariationsControllerSpec.scala]: should this be "validEstateVariationsRequestJson" ?

File: app/uk/gov/hmrc/trusts/controllers/EstateVariationsController.scala (hunk 1 above)
Comment (JoshuaH01) [reply app/uk/gov/hmrc/trusts/controllers/EstateVariationsController.scala]: done

File: test/uk/gov/hmrc/trusts/controllers/EstateVariationsControllerSpec.scala (hunk 3 above)
Comment (JoshuaH01) [reply test/uk/gov/hmrc/trusts/controllers/EstateVariationsControllerSpec.scala]: Yes it should be. complete.

File: test/uk/gov/hmrc/trusts/controllers/EstateVariationsControllerSpec.scala (hunk 4 above)
Comment (JoshuaH01) [reply test/uk/gov/hmrc/trusts/controllers/EstateVariationsControllerSpec.scala]: Yes it should be estate. complete.

File: app/uk/gov/hmrc/trusts/controllers/EstateBaseController.scala [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+import play.api.data.validation.ValidationError
+import play.api.libs.json._
+import play.api.mvc.{Request, Result}
+import uk.gov.hmrc.play.bootstrap.controller.BaseController
+import uk.gov.hmrc.trusts.utils.ErrorResponses._
+
+import scala.concurrent.Future
+
+class EstateBaseController extends BaseController {
```
Comment (adamconder) [inline app/uk/gov/hmrc/trusts/controllers/EstateBaseController.scala]: I don't think we need to create this EstateBaseController. We're duplicating a lot of functionality that is the same across estates and trusts.

File: app/uk/gov/hmrc/trusts/controllers/EstateVariationsController.scala [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+              request.body,
+              request.identifier,
+              response = Json.toJson(response)
+            )
+
+            Ok(Json.toJson(response))
+
+          } recover {
+
+            case InvalidCorrelationIdException =>
```
Comment (adamconder) [inline app/uk/gov/hmrc/trusts/controllers/EstateVariationsController.scala]: This case is not needed any more. It isn't unique to just this PR but we do not validate the CorrelationId in the HTTP request headers therefore cannot be returned an `InvalidCorrelationIdException`.

File: app/uk/gov/hmrc/trusts/controllers/GetTrustController.scala [hunk 7]
```diff
@@ -65,7 +65,7 @@ class GetTrustController @Inject()(identify: IdentifierAction,
 
         case _: ResourceNotFoundResponse.type =>
           auditService.auditErrorResponse(TrustAuditing.GET_TRUST, Json.obj("utr" -> utr), request.identifier, "Not Found received from DES.")
-          NotFound
+          InternalServerError
```
Comment (adamconder) [inline app/uk/gov/hmrc/trusts/controllers/GetTrustController.scala]: This should be returned to the old implementation. It should return a 404 NotFound.

File: app/uk/gov/hmrc/trusts/controllers/TrustVariationsController.scala [hunk 8]
```diff
@@ -19,32 +19,27 @@ package uk.gov.hmrc.trusts.controllers
...
                                     ) extends TrustsBaseController with ValidationUtil {
 
-  def variation() = identify.async(parse.json) {
+  def trustVariation() = identify.async(parse.json) {
     implicit request =>
 
-      val payload = request.body.toString
-
-      validationService.get(config.variationsApiSchema).validate[Variation](payload).fold(
+      request.body.validate[TrustVariation].fold(
```
Comment (adamconder) [inline app/uk/gov/hmrc/trusts/controllers/TrustVariationsController.scala]: You have removed the `validationService` and the call to `validationService.get(config.variationsApiSchema).validate[Variation](payload).fold()` due to the way the branch was created off the branch and merged back in.
The commits were squashed and this has resulted in the bad merge.

File: conf/estates.routes [hunk 9]
```diff
@@ -5,3 +5,5 @@ POST        /check                    uk.gov.hmrc.trusts.controllers.CheckEstate
 POST        /register                 uk.gov.hmrc.trusts.controllers.RegisterEstateController.registration
 
 GET         /:utr                     uk.gov.hmrc.trusts.controllers.GetEstateController.get(utr: String)
+
+POST        /estatevariation          uk.gov.hmrc.trusts.controllers.EstateVariationsController.estateVariation
```
Comment (adamconder) [inline conf/estates.routes]: The endpoint should just be `/variation`

File: conf/trusts.routes [hunk 10]
```diff
@@ -8,4 +8,4 @@ POST        /tax-enrolment/callback/subscriptions      uk.gov.hmrc.trusts.contro
 
 GET         /:utr                                      uk.gov.hmrc.trusts.controllers.GetTrustController.get(utr: String)
 
-POST        /variation                                 uk.gov.hmrc.trusts.controllers.VariationsController.variation
+POST        /trustvariation                            uk.gov.hmrc.trusts.controllers.TrustVariationsController.trustVariation
```
Comment (adamconder) [inline conf/trusts.routes]: The endpoint should just be `/variation`

File: test/uk/gov/hmrc/trusts/controllers/EstateVariationsControllerSpec.scala [hunk 11]
```diff
@@ -31,56 +29,55 @@ import uk.gov.hmrc.trusts.BaseSpec
...
+
   lazy val mockAuditService: AuditService = mock[AuditService]
 
-  val mockAuditConnector = mock[AuditConnector]
-  val mockConfig = mock[AppConfig]
+  val mockAuditConnector: AuditConnector = mock[AuditConnector]
+  val mockConfig: AppConfig = mock[AppConfig]
 
   val auditService = new AuditService(mockAuditConnector, mockConfig)
-  val validationService = new ValidationService()
```
Comment (adamconder) [inline test/uk/gov/hmrc/trusts/controllers/EstateVariationsControllerSpec.scala]: We need to add back in validationService.

File: test/uk/gov/hmrc/trusts/controllers/EstateVariationsControllerSpec.scala [hunk 12]
```diff
@@ -90,22 +87,21 @@ class VariationsControllerSpec extends BaseSpec with BeforeAndAfter {
...
           .thenReturn(Future.successful(VariationResponse(tvnResponse)))
 
         when(mockConfig.auditingEnabled).thenReturn(true)
-        when(mockConfig.variationsApiSchema).thenReturn(appConfig.variationsApiSchema)
 
-        val requestPayLoad = Json.parse(validVariationsRequestJson)
+        val requestPayLoad = Json.parse(validEstateVariationsRequestJson)
 
-        val SUT = new VariationsController(new FakeIdentifierAction(Organisation), mockConfig, mockDesService, auditService, validationService)
+        val SUT = new EstateVariationsController(new FakeIdentifierAction(Organisation), mockDesService, auditService)
```
Comment (adamconder) [inline test/uk/gov/hmrc/trusts/controllers/EstateVariationsControllerSpec.scala]: We need to add back in validationService mock.

File: test/uk/gov/hmrc/trusts/controllers/EstateVariationsControllerSpec.scala [hunk 13]
```diff
@@ -176,12 +173,12 @@ class VariationsControllerSpec extends BaseSpec with BeforeAndAfter {
 
       "input request fails business validation" in {
 
-        when(mockDesService.variation(any[Variation])(any[HeaderCarrier]))
+        when(mockDesService.estateVariation(any[EstateVariation])(any[HeaderCarrier]))
           .thenReturn(Future.failed(BadRequestException))
 
-        when(mockConfig.variationsApiSchema).thenReturn(appConfig.variationsApiSchema)
```
Comment (adamconder) [inline test/uk/gov/hmrc/trusts/controllers/EstateVariationsControllerSpec.scala]: Need to add this back into all tests?

File: test/uk/gov/hmrc/trusts/controllers/TrustVariationsControllerSpec.scala [hunk 14]
```diff
@@ -0,0 +1,332 @@
...
+import uk.gov.hmrc.trusts.config.AppConfig
+import uk.gov.hmrc.trusts.controllers.actions.FakeIdentifierAction
+import uk.gov.hmrc.trusts.exceptions._
+import uk.gov.hmrc.trusts.models.variation.{TrustVariation, VariationResponse}
+import uk.gov.hmrc.trusts.services.{AuditService, DesService}
+import uk.gov.hmrc.trusts.utils.Headers
+
+import scala.concurrent.Future
+
+class TrustVariationsControllerSpec extends BaseSpec with BeforeAndAfter {
```
Comment (adamconder) [inline test/uk/gov/hmrc/trusts/controllers/TrustVariationsControllerSpec.scala]: TrustsVariationsController specs do not reflect what is in master.
We need to add back in the validation service.

File: test/uk/gov/hmrc/trusts/connectors/DesConnectorSpec.scala (hunk 2 above)
Comment (JoshuaH01) [reply test/uk/gov/hmrc/trusts/connectors/DesConnectorSpec.scala]: Complete

File: app/uk/gov/hmrc/trusts/controllers/EstateBaseController.scala (hunk 5 above)
Comment (JoshuaH01) [reply app/uk/gov/hmrc/trusts/controllers/EstateBaseController.scala]: removed esatesbasecontroller and trustbasecontroller, merged them in to variationsbasecontroller

File: app/uk/gov/hmrc/trusts/controllers/EstateVariationsController.scala (hunk 6 above)
Comment (JoshuaH01) [reply app/uk/gov/hmrc/trusts/controllers/EstateVariationsController.scala]: Discussed with Darrel, decided keeping this case was best

File: app/uk/gov/hmrc/trusts/controllers/GetTrustController.scala (hunk 7 above)
Comment (JoshuaH01) [reply app/uk/gov/hmrc/trusts/controllers/GetTrustController.scala]: changed back to 404

File: app/uk/gov/hmrc/trusts/controllers/TrustVariationsController.scala (hunk 8 above)
Comment (JoshuaH01) [reply app/uk/gov/hmrc/trusts/controllers/TrustVariationsController.scala]: Added validation service to tests and implementations

File: conf/trusts.routes (hunk 10 above)
Comment (JoshuaH01) [reply conf/trusts.routes]: removed the trustvariation and changed to variation

File: test/uk/gov/hmrc/trusts/controllers/EstateVariationsControllerSpec.scala (hunk 12 above)
Comment (JoshuaH01) [reply test/uk/gov/hmrc/trusts/controllers/EstateVariationsControllerSpec.scala]: Fixed

File: test/uk/gov/hmrc/trusts/controllers/EstateVariationsControllerSpec.scala (hunk 13 above)
Comment (JoshuaH01) [reply test/uk/gov/hmrc/trusts/controllers/EstateVariationsControllerSpec.scala]: Complete

File: test/uk/gov/hmrc/trusts/controllers/TrustVariationsControllerSpec.scala (hunk 14 above)
Comment (JoshuaH01) [reply test/uk/gov/hmrc/trusts/controllers/TrustVariationsControllerSpec.scala]: complete

File: conf/estates.routes (hunk 9 above)
Comment (JoshuaH01) [reply conf/estates.routes]: Changed to variation

File: test/uk/gov/hmrc/trusts/controllers/EstateVariationsControllerSpec.scala (hunk 11 above)
Comment (JoshuaH01) [reply test/uk/gov/hmrc/trusts/controllers/EstateVariationsControllerSpec.scala]: Added

**Pull request merged**